File: packages/protocol/src/authoring.ts
import { z } from "zod";
import { beatWireSchema } from "./beat.js";
import { automationSourceSchema } from "./automation-source.js";
import { entityIdSchema, pitchSchema } from "./primitives.js";
import { effectRefSchema, instrumentRefSchema } from "./refs.js";
import { loopSpecSchema } from "./timeline.js";

export const noteSpecSchema = z.object({
  id: entityIdSchema.optional(),
  pitch: pitchSchema,
  start: beatWireSchema,
  duration: beatWireSchema,
  velocity: z.number().finite().min(0).max(1),
  offVelocity: z.number().finite().min(0).max(1).optional(),
  chance: z.number().finite().min(0).max(1).optional(),
  voice: z.number().int().nonnegative().optional(),
  tags: z.array(z.string()).optional(),
});
export type NoteSpec = z.infer<typeof noteSpecSchema>;

export const patternSpecSchema = z.object({
  id: entityIdSchema,
  name: z.string().optional(),
  lengthBeats: beatWireSchema,
  notes: z.array(noteSpecSchema),
  /** Independent Channel parts; leaves share the root loop period and may be shorter. */
  parts: z.array(z.object({ channelId: entityIdSchema, patternId: entityIdSchema })).min(1).max(256).optional(),
});
export type PatternSpec = z.infer<typeof patternSpecSchema>;

export const patternClipSpecSchema = z
  .object({
    id: entityIdSchema,
    patternId: entityIdSchema,
    trackId: entityIdSchema,
    startBeat: beatWireSchema,
    durationBeats: beatWireSchema.optional(),
    loopCount: z.number().int().min(1).optional(),
    lastBeat: beatWireSchema.optional(),
    transpose: z.number().int().optional(),
    velocityScale: z.number().finite().min(0).max(2).optional(),
    probability: z.number().finite().min(0).max(1).optional(),
    enabled: z.boolean().optional(),
  })
  .refine((c) => !(c.loopCount !== undefined && c.lastBeat !== undefined), {
    message: "loopCount and lastBeat are mutually exclusive",
  });
export type PatternClipSpec = z.infer<typeof patternClipSpecSchema>;

export const sampleClipSpecSchema = z.object({
  id: entityIdSchema,
  sampleId: entityIdSchema,
  trackId: entityIdSchema,
  startBeat: beatWireSchema,
  durationBeats: beatWireSchema.optional(),
  gain: z.number().finite().min(0).max(2).optional(),
  pan: z.number().finite().min(-1).max(1).optional(),
  rate: z.number().finite().min(0.25).max(4).optional(),
  loop: loopSpecSchema.optional(),
  tempoSync: z.enum(["off", "stretch", "repitch"]).optional(),
  stretchAlgorithm: z.string().min(1).optional(),
  enabled: z.boolean().optional(),
});
export type SampleClipSpec = z.infer<typeof sampleClipSpecSchema>;

export const channelSpecSchema = z.object({
  id: entityIdSchema,
  name: z.string().optional(),
  instrument: instrumentRefSchema,
  effectChain: z.array(effectRefSchema),
  level: z.number().finite().min(0).max(2),
  pan: z.number().finite().min(-1).max(1),
  swing: z.number().finite().min(0).max(1).optional(),
  mixerChannelId: entityIdSchema,
  mute: z.boolean().optional(),
  solo: z.boolean().optional(),
});
export type ChannelSpec = z.infer<typeof channelSpecSchema>;

export const sendSpecSchema = z.object({
  destinationId: entityIdSchema,
  ratio: z.number().finite().min(0).max(1),
  preFader: z.boolean().optional(),
  sidechain: z.boolean().optional(),
});
export type SendSpec = z.infer<typeof sendSpecSchema>;

export const mixerChannelSpecSchema = z.object({
  id: entityIdSchema,
  name: z.string().optional(),
  level: z.number().finite().min(0).max(2),
  balance: z.number().finite().min(-1).max(1),
  masterSendRatio: z.number().finite().min(0).max(1).optional(),
  inserts: z.array(effectRefSchema),
  sends: z.array(sendSpecSchema),
  mute: z.boolean().optional(),
  solo: z.boolean().optional(),
});
export type MixerChannelSpec = z.infer<typeof mixerChannelSpecSchema>;

export const automationLaneSpecSchema = z.object({
  id: entityIdSchema,
  target: z.object({ entityId: entityIdSchema, parameterId: z.string().min(1), scope: z.enum(["plugin", "effectHost"]).optional() }),
  source: automationSourceSchema,
  combine: z.enum(["replace", "add", "multiply", "max"]).optional(),
  playback: z.enum(["global", "playlist"]).optional(),
  loop: loopSpecSchema.optional(),
  lastBeat: beatWireSchema.optional(),
});
export type AutomationLaneSpec = z.infer<typeof automationLaneSpecSchema>;

export const automationClipSpecSchema = z.object({
  id: entityIdSchema,
  laneId: entityIdSchema,
  trackId: entityIdSchema,
  startBeat: beatWireSchema,
  durationBeats: beatWireSchema.optional(),
  enabled: z.boolean().optional(),
});
export type AutomationClipSpec = z.infer<typeof automationClipSpecSchema>;

File: crates/core/src/wire/authoring.rs
//! Authoring wire types: notes, patterns, clips, channels, mixer and
//! automation lanes (04-api-contracts.md §Authoring interfaces).

use serde::{Deserialize, Serialize};

use super::automation::AutomationSourceSpec;
use super::basic::{EffectRef, InstrumentRef, LoopSpec};
use super::EntityId;
use crate::beat::Beat;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NoteSpec {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<EntityId>,
    pub pitch: u8,
    pub start: Beat,
    pub duration: Beat,
    pub velocity: f64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub off_velocity: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub chance: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub voice: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatternSpec {
    pub id: EntityId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub length_beats: Beat,
    pub notes: Vec<NoteSpec>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub parts: Option<Vec<PatternPartSpec>>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatternPartSpec {
    pub channel_id: EntityId,
    pub pattern_id: EntityId,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatternClipSpec {
    pub id: EntityId,
    pub pattern_id: EntityId,
    pub track_id: EntityId,
    pub start_beat: Beat,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub duration_beats: Option<Beat>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub loop_count: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_beat: Option<Beat>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub transpose: Option<i32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub velocity_scale: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub probability: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub enabled: Option<bool>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TempoSync {
    Off,
    Stretch,
    Repitch,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SampleClipSpec {
    pub id: EntityId,
    pub sample_id: EntityId,
    pub track_id: EntityId,
    pub start_beat: Beat,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub duration_beats: Option<Beat>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub gain: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pan: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rate: Option<f64>,
    #[serde(default, rename = "loop", skip_serializing_if = "Option::is_none")]
    pub loop_spec: Option<LoopSpec>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tempo_sync: Option<TempoSync>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub stretch_algorithm: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub enabled: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChannelSpec {
    pub id: EntityId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub instrument: InstrumentRef,
    pub effect_chain: Vec<EffectRef>,
    pub level: f64,
    pub pan: f64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub swing: Option<f64>,
    pub mixer_channel_id: EntityId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mute: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub solo: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SendSpec {
    pub destination_id: EntityId,
    pub ratio: f64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pre_fader: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sidechain: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MixerChannelSpec {
    pub id: EntityId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub level: f64,
    pub balance: f64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub master_send_ratio: Option<f64>,
    pub inserts: Vec<EffectRef>,
    pub sends: Vec<SendSpec>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mute: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub solo: Option<bool>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AutomationCombine {
    Replace,
    Add,
    Multiply,
    Max,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AutomationTarget {
    pub entity_id: EntityId,
    pub parameter_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub scope: Option<ParameterScope>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ParameterScope {
    Plugin,
    EffectHost,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AutomationPlayback {
    Global,
    Playlist,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AutomationLaneSpec {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub playback: Option<AutomationPlayback>,
    pub id: EntityId,
    pub target: AutomationTarget,
    pub source: AutomationSourceSpec,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub combine: Option<AutomationCombine>,
    #[serde(default, rename = "loop", skip_serializing_if = "Option::is_none")]
    pub loop_spec: Option<LoopSpec>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_beat: Option<Beat>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AutomationClipSpec {
    pub id: EntityId,
    pub lane_id: EntityId,
    pub track_id: EntityId,
    pub start_beat: Beat,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub duration_beats: Option<Beat>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub enabled: Option<bool>,
}
